Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,213 @@
+* "hosting components" - shorthand for .NET Core hosting components. Typically refers to `hostfxr` and `hostpolicy`. Sometimes also referred to simply as "host".
+
+### Initialize host context
+
+All the "initialize" functions will
+* Process the `.runtimeconfig.json`
+* Resolve framework references and find actual frameworks
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the TPA, native search paths and other artifacts.
+
+The functions will NOT load the CoreCLR runtime. They just prepare everything to the point where it can be loaded.
+
+The functions return a handle to a new host context:
+* The handle must be closed via `hostfxr_shutdown`.
+* The handle is not thread safe - the consumer should only call functions on it from one thread at a time.
+
+The `hostfxr` will also track active runtime in the process. Due to limitations (and to simplify implementation) this tracking will actually not look at the actual `coreclr` module (or try to communicate with the runtime in any way). Instead `hostfxr` itself will track the host context initialization. The first host context initialization in the process will represent the "loaded runtime". It is only possible to have one "loaded runtime" in the process. Any subsequent host context initialization will just "attach" to the "loaded runtime" instead of creating a new one.
+
+*Note: It is technically possible to initialize host context, and never call any "run" method on it, which would mean the runtime would not get loaded. Trying to initialize a second host context while the first one didn't start the runtime yet will for now be considered illegal. It's a pending issue to solve this problem, as COM activation will probably need to be able to safely initialize/run from multiple threads at a time.*
+
+``` C++
+using hostfxr_handle = void *;
+
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;
+    const char_t * dotnet_root;
+};
+```
+
+The `hostfxr_initialize_parameters` structure stores parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the hosting components. It's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C++
```

Comment:
This is an export right? If so please update the language to be C or prefix the signature with `extern "C"`. It also might be a good place to comment on calling convention or link to the ABI contracts the hosting API follows.

Reply:
I was trying to avoid to write the full blown header file in the spec... I'll change the language to C, but I think including full calling convention specification in the spec feels abit over the top. (I agree it's important, but not in the spec I think)

Reply:
Added a note about exports and that they follow the same style as the other exports in `hostfxr`.